docs/proposals/20190704-AutomatedUpgrades.md

```diff
@@ -0,0 +1,166 @@
+This is part of the solution to remove the need for `ROOT-A` and `ROOT-B` partitions.
+It also allows us to dedicate the owned block device to data that can be reproduced (e.g. containerd data, kubelet data, etcd data, CNI data, etc.).
+
+### Using `squashfs` for the `rootfs`
+
+For performance reasons, and a smaller footprint, we can create our `rootfs` as a `squashfs` image.
+The way in which we can run the `rootfs` is by mounting it using a loopback device.
+Once mounted, we have will perform a `switch_root` into the mounted `rootfs`.
+
+> Note: A benefit of `squashfs` is that it allows us to retain our read only `rootfs` without having to mount it as such.
+
+### Slimming down `rootfs`
+
+To slim down the size of our rootfs, I propose that we remove Kubernetes tarballs entirely and let the `kubelet` download them.
+This is a shift from our immutability goals, but in this case we can embrace configurability at the Kubernetes layer.
+The disadvantage to this is that this introduces a number a variables in the number of things that can go wrong.
+Having "approved" tarballs packaged with Talos allows us to run conformance tests against exactly what users will use.
+That being said, I think this is overall a better direction for Talos.
+
+### A New Partition Scheme
+
+With the move to run the `rootfs` in memory, we will remove the need for `ROOT-A` and `ROOT-B` partitions.
+In the case of cloud installs, or where users may want to boot without PXE/iPXE, we will require a boot partition.
+In all cases we require a data partition.
+
+I also propose we rename the `DATA` partition to `EPHEMERAL` to make it clear that this partition should not be used to store application data.
+Our suggestion to users who do want persistent storage on the host is to mount another disk and to use that.
+Talos will make the guarantee to only manage the "owned" block device.
+
+### Performing Upgrades in a Container
+
+The move to running `rootfs` in memory, and thus making the block device ephemeral, and the split of `init` into `init` and `machined`, allows us to take an approach to upgrades that is clean, and relatively low risk.
+The workflow will look roughly like the following:
+
+- an upgrade request is received by `osd` and proxied to `machined`
+- `machined` stops `system` and `k8s.io` services
+- `machined` unmounts all mounts in `/var`
+- `machined` runs the specified installer container
+
+### Automated Upgrades
+
+All of the aformentioned improvements help in the higher goal of automated upgrades.
+Once the above is implemented, upgrades should be simple and self-contained to a machine.
+By running upgrades in containers, we can have very simple operator implementation that can rollout an upgrade across a cluster by simply:
+
+- checking for a new tag
+- orchestrating cluster-wide rollout
+
+The operator will initially run upgrades only, but it will also provide a basis for a number of future automation opportunities.
```

Comment:
Perhaps "coordinate upgrades" rather than "run upgrades" would be more accurate?

Reply:
This is a good point. I don’t we are entirely sure or we haven’t discussed at depth how upgrades would work from the operator point of view. I guess there are two ways. One is the way you are suggesting. If I understand this suggestion it would work by having each individual node invoke their own upgrade but perhaps getting a lock to do so. The alternative is that the operator invoking the node’s API and manages the scheduling internally. Do you see benefits of one over the other? I plan on writing a separate proposal, or maybe even expanding this one, about how this would work so good if we start the discussion now.

Reply:
Perhaps I am being too pedantic, but both of those are examples of the operator "coordinating" the upgrades, by my meaning.  I don't think the former is a good idea, topologically.  That is, I don't think the talos components should reach into kubernetes to acquire the lock.  
The flow should, in my opinion flow from the operator to talos, only.  Thus, the scheduling, locking, etc, would all be under the strict purview of the operator.

Basically, in the ideal case, an operator should act as a human, using the same tools and APIs as a human, just in an automated fashion.

Reply:
It's the problem with text reviews.  I start focusing on the English rather than the engineering intention.

Reply:
> Basically, in the ideal case, an operator should act as a human, using the same tools and APIs as a human, just in an automated fashion.

Agree with this 100%. I will add some more to this proposal around how the operator will work. 

> It's the problem with text reviews. I start focusing on the English rather than the engineering intention.

My bad English probably doesn’t help.

Reply:
Added a section that hopefully capture this.